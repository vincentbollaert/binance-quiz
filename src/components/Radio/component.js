import React from 'react'
import { bool, string, number } from 'prop-types'
import styled from 'styled-components'
import { UNIT_LG, SELECTIVE_YELLOW, TRANSITION, GRANITE_GRAY } from '../../styles'
import { CN_ANSWER } from '../../pages/Home/shared'

export const STYLE_RADIO_SIZE_PX = 24
export const CN_RADIO = 'radio'

export const Wrap = styled.div`
  display: flex;
  align-items: center;
  margin-right: ${UNIT_LG};
`

export const RadioWrap = styled.div`
  position: relative;
  width: ${STYLE_RADIO_SIZE_PX}px;
  height: ${STYLE_RADIO_SIZE_PX}px;

  ${props => !props.isDisabled && `
    .${CN_ANSWER}:hover & {
      .${CN_RADIO} {
        border-color: ${SELECTIVE_YELLOW};

        &::before {
          visibility: visible;
          background-color: ${SELECTIVE_YELLOW};
        };
      };
    };
  `};
`

export const RadioField = styled.input`
  display: none;

  &:checked ~ {
    .${CN_RADIO} {
      border-color: ${props => props.accentColor};

      &::before {
        visibility: visible;
        width: ${STYLE_RADIO_SIZE_PX / 2}px;
        height: ${STYLE_RADIO_SIZE_PX / 2}px;
        background-color: ${props => props.accentColor};
      };
    };
  };
`

export const RadioLabel = styled.label`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
`

export const RadioStyled = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 2px solid ${GRANITE_GRAY};
  width: 100%;
  height: 100%;
  border-radius: 50%;

  &::before {
    content: '';
    position: absolute;
    visibility: hidden;
    height: ${STYLE_RADIO_SIZE_PX / 4}px;
    width: ${STYLE_RADIO_SIZE_PX / 4}px;
    border-radius: 50%;
    background-color: ${GRANITE_GRAY};
    transition: width ${TRANSITION}, height ${TRANSITION}, background-color ${TRANSITION};
    pointer-events: none;
  }
`

const Radio = ({ isQuestionFinished, accentColor, id, className }) => (
  <Wrap className={className}>
    <RadioWrap accentColor={accentColor} isDisabled={isQuestionFinished}>
      <RadioField
        accentColor={accentColor}
        checked={accentColor !== undefined}
        type="radio"
        id={id}
        name={id}
        readOnly
      />
      <RadioLabel htmlFor={id} />
      <RadioStyled className={CN_RADIO} />
    </RadioWrap>
  </Wrap>
)

Radio.propTypes = {
  isQuestionFinished: bool.isRequired,
  accentColor: string,
  id: number.isRequired,
  className: string,
}

Radio.defaultProps = {
  accentColor: undefined,
  className: '',
}

export default Radio
