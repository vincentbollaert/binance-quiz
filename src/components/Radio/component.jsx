import React from 'react'
import { bool, string, number } from 'prop-types'
import styled from 'styled-components'
import { UNIT_LG, SELECTIVE_YELLOW } from '../../styles'

const SIZE = 24
const BG = '#636363'
export const RADIO_STYLED_CLASSNAME = 'radio-styled'

export const Wrap = styled.div`
  display: flex;
  align-items: center;
  margin-right: ${UNIT_LG};
`

export const RadioWrap = styled.div`
  position: relative;
  width: ${SIZE}px;
  height: ${SIZE}px;

  ${props => !props.isDisabled && `
    .sdsds:hover & {
      .${RADIO_STYLED_CLASSNAME} {
        border-color: ${props.accentColor};

        &::before {
          visibility: visible;
          background-color: ${props.accentColor};
        }
      }
    }
  `}
`

export const RadioField = styled.input`
  display: none;

  &:checked ~ {
    .${RADIO_STYLED_CLASSNAME} {
      border-color: ${props => props.accentColor};

      &::before {
        visibility: visible;
        width: ${SIZE / 2}px;
        height: ${SIZE / 2}px;
        background-color: ${props => props.accentColor};
      }
    }
  }
`

export const RadioLabel = styled.label`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
`

export const RadioStyled = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  border: 2px solid ${BG};
  width: 100%;
  height: 100%;
  border-radius: 50%;

  &::before {
    content: '';
    position: absolute;
    visibility: hidden;
    height: ${SIZE / 4}px;
    width: ${SIZE / 4}px;
    border-radius: 50%;
    background-color: ${BG};
    transition: width 0.1s ease-out, height 0.1s ease-out, background-color 0.1s ease-out;
    pointer-events: none;
  }
`

class Radio extends React.PureComponent {
  render() {
    const {
      isQuizComplete,
      isDisabled,
      checked,
      accentColor,
      id,
      name,
      className,
    } = this.props
    const accentColorConditional = isQuizComplete ? accentColor : SELECTIVE_YELLOW

    return (
      <Wrap className={className}>
        <RadioWrap accentColor={accentColorConditional} isDisabled={isDisabled}>
          <RadioField accentColor={accentColorConditional} checked={checked} type="radio" id={id} name={name} readOnly />
          <RadioLabel htmlFor={id} />
          <RadioStyled className={RADIO_STYLED_CLASSNAME} />
        </RadioWrap>
      </Wrap>
    )
  }
}

Radio.propTypes = {
  isQuizComplete: bool.isRequired,
  isDisabled: bool,
  checked: bool,
  accentColor: string.isRequired,
  id: number.isRequired,
  name: number,
  className: string,
}

Radio.defaultProps = {
  isDisabled: false,
  checked: undefined,
  name: undefined,
  className: '',
}

export default Radio
