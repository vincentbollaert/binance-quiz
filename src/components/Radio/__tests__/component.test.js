import '@testing-library/jest-dom'
import React from 'react'
import { render } from '@testing-library/react'
import Radio from '../component'

describe('Radio', () => {
  const props = {
    isQuestionFinished: false,
    id: '1',
    accentColor: undefined,
  }
  test('matches snapshot', () => {
    const { container } = render(<Radio {...props} />)
    expect(container.firstChild).toMatchSnapshot()
  })
  test('renders elements', () => {
    const { container } = render(<Radio {...props} isShow />)
    expect(container.querySelector('input')).not.toBeChecked()
  })
  test('renders elements', () => {
    const { container } = render(<Radio {...props} isShow accentColor="red" />)
    expect(container.querySelector('input')).toBeChecked()
  })
})
