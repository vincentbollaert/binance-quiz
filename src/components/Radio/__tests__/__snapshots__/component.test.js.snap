// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio matches snapshot 1`] = `
<div
  class="sc-AxjAm jCPDH"
>
  <div
    class="sc-AxirZ hgttxX"
  >
    <input
      class="sc-AxiKw cpYWwn"
      id="1"
      name="1"
      readonly=""
      type="radio"
    />
    <label
      class="sc-AxhCb kPThuO"
      for="1"
    />
    <div
      class="sc-AxhUy kztugP radio"
    />
  </div>
</div>
`;
